### PR TITLE
Add support for generating a certificate signing request to genssl.

### LINKED ISSUE #1299: SSL Tool - Create CSR
It would be rather nice for the ssl tool to create a CSR to get a SSL Certificate with from like Comodo

### PR DESCRIPTION
**Pull request by SaberUK**
I have tested it locally on my machine and the output seems fine but it will require someone who knows more about CSRs (@Casey952 maybe?) to test further.

Fixes #1299

### CODE REVIEW

File: tools/genssl [hunk 1]
```diff
@@ -120,6 +120,7 @@ __GNUTLS_END__
 	$status ||= system "$certtool --generate-privkey $sec_param --outfile key.pem";
 	$status ||= system "$certtool --generate-self-signed --load-privkey key.pem --outfile cert.pem --template $tmp";
 	$status ||= system "$certtool --generate-dh-params $sec_param --outfile dhparams.pem";
+	$status ||= system "$certtool --generate-request $sec_param --load-privkey key.pem --outfile csr.pem --template $tmp";
```
Comment (SaberUK) [inline tools/genssl]: I'm not sure whether $sec_param is required here. What do you think @attilamolnar?

File: tools/genssl [hunk 2]
```diff
@@ -131,9 +132,12 @@ $organization
 $unit
 $common_name
 $email
+
+
```
Comment (SaberUK) [inline tools/genssl]: This white space is intentional btw.

**Comment by Casey952** [comment]
This is from Charybdis genssl

```
#!/bin/sh
prefix="@prefix@"
exec_prefix="@exec_prefix@"
sysconfdir="@sysconfdir@"

echo "Generating private key and CSR... "
openssl req -new -newkey rsa:4096 -nodes -sha512 -out "${sysconfdir}"/ssl.csr -keyout "${sysconfdir}"/ssl.key

echo "Self-signing certificate..."
openssl x509 -req -sha512 -days 365 -in "${sysconfdir}"/ssl.csr -signkey "${sysconfdir}"/ssl.key -out "${sysconfdir}"/ssl.pem

echo "Generating Diffie-Hellman file for secure SSL/TLS negotiation .. "
openssl dhparam -out "${sysconfdir}"/dh.pem 2048

# If sysconfdir is relative to prefix, make the path relative. I.e.,
# prefix=/usr and sysconfdir=/etc -> relative_sysconfdir=/etc,
# prefix=/home/binki/chary and sysconfdir=/home/binki/chary/etc ->
# relative_sysconfdir=etc
relative_sysconfdir="${sysconfdir#${prefix%/}/}"
relative_sysconfdir="${relative_sysconfdir%/}"

cat <<EOF


Now change these lines in the IRCd config file:

    ssl_private_key = "${relative_sysconfdir}/ssl.key";
    ssl_cert = "${relative_sysconfdir}/ssl.pem";
    ssl_dh_params = "${relative_sysconfdir}/dh.pem";

If you want to get your certificate signed by a certificate authority,
submit the ssl.csr file to your CA, then replace ssl.pem with the
certificate returned to you. You may need to include your CA's
intermediate certificates in signing order.

Enjoy using ssl.
EOF

```

File: tools/genssl (hunk 1 above)
Comment (attilamolnar) [inline tools/genssl]: It's not required.

Comment (SaberUK) [inline tools/genssl]: I've removed it.

**Pull request merged**